Sync Lock Explicit Leader

ywwg opened the pull request:
Enable sync lock explicit leader selection.  This change builds on #3695 and will be updated with changes that happen there.

This is a major change to the Sync Lock code that includes a bunch of cleanup and rationalization. Hopefully the logic is a little easier to follow now. This change adds a "Leader" button to the UI. This button is lit to indicate which deck is the Leader.  

**testing note! You must use Late Nite Classic to see the new Leader button.  This is very temp UI for testing.** Orange means that Mixxx picked that deck automatically.  That setting will bounce around as you load new tracks.  Clicking the button turns it red if you want an explicit master.

Things that need testing by hand:

* situations with beatgrid changes.  Loading new tracks, reanalyzing tracks, etc.  https://bugs.launchpad.net/mixxx/+bug/1808698 esp
* test out explicit leader! This means that the designated deck will have its bpm preferred above others.  This is really only noticeable when there are beatmap tracks.
* using one deck to scratch another is gone. It was too hard to make work with all the edge cases.
* Try letting an explicit leader hit the end of a track.  Is that the expected behavior for you?
* Just really beat it up.

ywwg committed:
Sync Lock: Fix leader button in latenight (very temp graphic)

ywwg committed:
Sync Lock: remove debugging stuff

ywwg committed:
Sync Lock: Minor refactoring.

This refactoring sets up a major major patching coming in for sync lock! Get hyped.
This is mostly a noop PR, except for the fix when playing in reverse.

ywwg committed:
Sync Lock: make sure there's always a sync leader

also remove obsolete tests

ywwg committed:
Sync Lock: Add orange state for sync leader button

ywwg committed:
Sync Lock: Fix some broken tests

daschuer commented:
> using one deck to scratch another is gone. It was too hard to make work with all the edge cases.

For my understanding this was the main use case for the explicit leader mode.
An other useful thing is to have a fixed internal/external clock. Will this be also part of this PR?  

What is the main use case for making a Deck explicit leader?  I cannot really think of one right now.

ywwg commented:
If I'm playing two beatmapped tracks, right now Mixxx doesn't know which one should drive the bpm, so currently it just locks the rate to the most recent value.  So the main use-case for this change is allowing the user to determine which track is "in charge" of the bpm when there are beatmapped tracks involved. Mixxx can either pick one or the other to drive the bpm, or the user can. (If we add UI for the Internal Clock, we could also lock the bpm to a constant value.)

I can work on restoring make-decks-scratch functionality in a future PR, but I need people to help me think through the edge cases (what happens when the master deck stops?  Everything stops?) etc. This change does not make that work impossible.

daschuer commented:
That sounds reasonable, however you have described the solution, not the use case.

I am a bit concerned that we get a good balance between feature and clutter in the GUI. We also need to have use cases in place where we can adjust the edge cases to.

To be honest I can't think of a real use case for playing two tracks in parallel at center Crossfader all the time. 

What I can think of to have a short section or loop in a track, that I want to add to the main track, without hearing jouling or something due to temporarily using a fixed clock. This is currently an issue that we should fix.

What else do we want?

ywwg commented:
Playing multiple tracks at the same time is necessary for mixing one track into another, usually over the course of 32 or 64 (or more) bars. With the current code, the sync lock picks a single bpm and locks both tracks to that.  This means the beatmap track is pushed and pulled around to remain at a constant bpm, which is not always desireable. This change allows the user to either pick the variable bpm, or a constant tempo.

This feature exists in Traktor and pioneer, and many controllers have a "master" button.  I thought we had already decided that explicit leader was a feature we wanted. (https://github.com/mixxxdj/mixxx/pull/2376)

New behavior aside, the code is much cleaner now.  There was a lot of hacky logic, and a lot of error-prone signal circles.  The new code is easier to follow now: enginesync.cpp contains all the logic for deciding which decks should be master, and synccontrol sends requests and notifications of state changes, and receives synclock bpm updates.

ywwg commented:
Even if we didn't include the new ui elements that I thought we had improved, this code is also still better because Mixxx automatically chooses a sync leader. Therefore if I'm playing a beatmapped song and bring something else in, Mixxx will keep respecting the variable bpm instead of suddenly locking to the current rate.  The user doesn't have to use the explicit leader button if they don't want to. But if they do want to be specific about which bpm they want, they can push the button.

Be-ing commented:
> if I'm playing a beatmapped song and bring something else in, Mixxx will keep respecting the variable bpm instead of suddenly locking to the current rate.

:100: 
The current behavior doesn't make a lot of sense and is quite awkward and unexpected.

ywwg committed:
Sync Lock: When disabling explicit sync, make sure we pick a soft master if there are two playing sync decks

ywwg committed:
Sync Lock: Remove unneeded end-of-track code

ronso0 commented:
> I am a bit concerned that we get a good balance between feature and clutter in the GUI

As with all advanced controls we can have an optional Sync controls section, like the beatgrid controls or vinyl control, especially for [Zulip: halve/double sync when the base/leader BPM changes significantly](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/sync.20lock.20edge.20cases)

daschuer commented:
> I thought we had already decided that explicit leader was a feature we wanted. (#2376)

Sorry I gave the impression I am in doubt with this. This feature is very welcome. And Thank you for picking it up. 

We need "just" to consider all use cases, to consider the edge cases correctly. This is also very helpful when considering the new analyzer features. 

> This means the beatmap track is pushed and pulled around to remain at a constant bpm, which is not always desirable.

Does this describe the jitter issue we have with the non const beat map? https://github.com/mixxxdj/mixxx/pull/3626 aims to fix this. 

In addition we have two types of tracks. Electronic track with intentional change and hand made tracks from the 80s or such. How does you used case works with them?

Holzhaus commented:
> Does this describe the jitter issue we have with the non const beat map? #3626 aims to fix this.

No. There are non-const tracks that are intentionally non-const  (either accelerando/ritardando or sudden tempo changes on a drop). Right now, it's not possible to use sync with such tracks properly because the tempo changes are ignored after enabling sync.

daschuer commented:
> There are non-const tracks that are intentionally non-const (either accelerando/ritardando or sudden tempo changes on a drop). 

Yes, that are the interesting tracks here. How does your use case work with them? 

> Right now, it's not possible to use sync with such tracks properly because the tempo changes are ignored after enabling sync.

It is ignored "only" if a second track is playing. This works OK because we can assume that a const tempo is the most common case and this feature avoids that the const track is yowling during a transition from one to another track. 
The result is a proper synced transition. 

When does this fail?

Holzhaus commented:
For example: two electronic tracks, track 1 with a sudden tempo change.

If I want to play both in sync and replace the high hats from track 1 with the ones from track 2, I currently can't do that.

daschuer commented:
Is that really a use case to play both side by side? I don't think that's a case.
I would normally loop the the track with the high heads. In this case the crowd dances to a track and I can decorate it with loops from other tracks.

Can you confirm it?

Holzhaus commented:
I don't think it matters if you loop the hihat track or not. In any case the tempo change from the other track won't be applied while synced.

Holzhaus commented:
By the way, I think we need this feature anyway for setting a midi clock input as explicit leader, so that track follow tempo changes from an external drum machine.

daschuer commented:
> I don't think it matters if you loop the hihat track or not.

I am not sure if it matters for the code. At least it matters to understand the use case. 

I think we can sort out the case to play two non const tracks in sync.





 In any case the tempo change from the other track won't be applied while synced.

ywwg commented:
> Is that really a use case to play both side by side? I don't think that's a case.

absolutely. If I'm mixing two disco tracks together, they will both have extremely variable bpms.  Furthermore, being able to mix two tracks together during a tempo change sounds really amazing.  Also note, the tracks don't both have to be playing out to master mix -- even if one track is gained down to zero and only audible in pfl, currently the sync will lock at an arbitrary bpm. So this fixes that case as well. 


I spun a whole set with this last night and it was great!

daschuer commented:
> If I'm mixing two disco tracks together, 

For my understanding this is the case where the const clock is the best solution. I am quite happy with this, because the youling of both tracks is limited during a transition from on to the other. 

> ...even if one track is gained down to zero and only audible in pfl, currently the sync will lock at an arbitrary bpm. So this fixes that case as well.

I think this can be considered as bug. I can't think of a case where a playing disco track should be locked to const while cuing a second track in. 

> Furthermore, being able to mix two tracks together during a tempo change sounds really amazing. 

Interesting. Can you share the tracks you have used for this?

ywwg commented:
> I think this can be considered as bug.

that's part of my point, this change also fixes that bug.  

>  const clock is the best solution

but this is weird because a track that had been variable is suddenly locked and it might cause audible artifacts as it changes speed to lock into place.  If you want a constant clock you can set the Internal Clock as the Explicit Leader.  This is exactly how Traktor does it btw :).

Be-ing commented:
IMO locking a variable track to a constant tempo is only appropriate if it is being synced to a constant tempo track which is the sync leader.

daschuer commented:
>> const clock is the best solution

> but this is weird because a track that had been variable is suddenly locked and it might cause audible artifacts as it changes speed to lock into place. 

I use non const beat maps since long and i did not experience artifacts. In almost all cases one of the tracks is const, so it irons only the beats of the non const track, which is a desired behavior.  

> If you want a constant clock you can set the Internal Clock as the Explicit Leader. This is exactly how Traktor does it btw :).

Do you propose to remove this implicit feature? I am still convinced that this is the best working solution for 99% of the cases, and a no brainier for the user.

daschuer commented:
> IMO locking a variable track to a constant tempo is only appropriate if it is being synced to a constant tempo track which is the sync leader.

@Be-ing: Since you use const beat grids luckily this is always true for you. 

If you have a transition of two disco tracks you cant really predict if there are slight tempo changes in the outro or intro. In addition that is even worse due to the jitter issue that will be fixed after merging https://github.com/mixxxdj/mixxx/pull/3626. Our detector also has sometimes issues to detect a stable BPM during intro or outros. I don't know my tracks that well that I know whether a track is affected. I don't know if I should select the in-going our out-going as leader. For this case the current locking strategy works perfectly.

daschuer commented:
@ywwg I am really curious to to reproduce your changing tempo transition. Can you share the used track titles?

daschuer commented:
@JoergAtGithub I am also interested to learn about your use cases. Do you use const beat grids or Beat maps? When do you miss the explicit master feature hard?

Holzhaus commented:
> I don't know if I should select the in-going our out-going as leader. For this case the current locking strategy works perfectly.

Then don't set an explicit leader. In that case the internal clock will be leader and sync works as before.

Are you saying we shouldn't add explicit leader support? Because I otherwise I don't understand what we are discussing here.

JoergAtGithub commented:
I like the Traktor behavior, where it's always clear which deck is the leader and I wish to have DVS control to this leader.
Especially for three deck transitions, it's good to know who leads and who follows.
I also like to play the same track on two decks in different flavors. These decks must be always in sync to sound like one track.
A single reference clock is also needed for main effects like echo, look at all the Pioneer gear, which has only one effects unit with beat related controls.

daschuer commented:
> Then don't set an explicit leader. In that case the internal clock will be leader and sync works as before.

Right. I am not arguing against the explicit leader feature. 

I have only the impression that we have a mixture of bugs and missing features, that we need to sort out to be finally better than Traktor and well prepared for the new beat grid. 

> Are you saying we shouldn't add explicit leader support? Because I otherwise I don't understand what we are discussing here.

I think I need to start every post with "I am happy that we are about to reintroduce the explicit sync feature" which is the case, .. really. 

I am currently quite happy with the current behavior for the general case. During the recent discussion I don't understand why it is considered as bad in general. 

For me these use cases are most important. 
1. Transition from one to another track using a const BPM even if both tracks have a non const beat map.
2. Decorating a track with loops or samples form on other track. Here the decorated track shall be the master
3. Scratching 2.
4. Fixed tempo by a dedicated clock (intern or extern) 

Are these the common use case for most other users as well?

daschuer commented:
@JoergAtGithub Ah, so I guess your used cases match my 2. and 4. right? Or is your case not yet covered? 

Are you use the constant setting in the analyzer? 

> I also like to play the same track on two decks in different flavors. These decks must be always in sync to sound like one track.

That should already work. Do you experience an issue with it? 

> Especially for three deck transitions, it's good to know who leads and who follows.

Is this currently really an issue in Mixxx? If you play currently three tracks with sync enabled, all tracks stick to the same const BPM. All tracks follow all rate sliders as if it is one. So you should have full control. 

There is an issue that the rate slider of the controller are out of sync and require soft take over. But that is an issue in every solution which automatically changes the rate.

daschuer commented:
@ywwg Can you rebase this branch on current main? It contains still the old commits of the rebased branch that is merged now. It would be also nice if you can fix the clazy warnings in the original commits during the rebase.

JoergAtGithub commented:
> @JoergAtGithub Ah, so I guess your used cases match my 2. and 4. right?

Yes

> Are you use the constant setting in the analyzer?

Yes

> Is this currently really an issue in Mixxx? If you play currently three tracks with sync enabled, all tracks stick to the same const BPM.

IMHO the current logic is really difficult to understand. It could be so much easier, if there would be a dedicated GUI control for the one and only leader.

> All tracks follow all rate sliders as if it is one. 

I don't use the software sliders.

Holzhaus commented:
I just tested this and it works fine IMHO.

Non-const track: Robert Ouimet - Rainbows Refunked: https://www.youtube.com/watch?v=ByY0hMIyc_4

Before, it was more basically impossible to use sync with this track because BPM fluctuates between ~115.5 and ~118.2 BPM. Now it actually works. The other track is jouling, but it is hardly noticable if you have keylock on. Certainly far better than the track going out of sync all the time as was the case before.

ywwg commented:
(what does "jouling" mean?)

daschuer commented:
> IMHO the current logic is really difficult to understand. It could be so much easier, if there would be a dedicated GUI control for the one and only leader.

Ah Ok. With a const beat grids, there is zero difference between the current behavior and an explicit master introduced here from the sync itself. The control and visualization is however different. So I probably starting to understand your issue. 

Please correct me If I am  wrong: 

When you enable sync on two decks, the rate slider of the first deck is is still snapped in while the rate slider of the second deck is out of sync. With an explicit master, you have an indicator which rate slider is the one that instantly controls the tempo and which need to be snapped in. 

Is this missing indicator part of your issue? 

Most controllers have only a single sync button that is lit when the deck is either leader or follower. How is the explicit leader indicated?

Holzhaus commented:
I don't know the correct English term for it. "Wow and flutter"? Like when you had and old cassette tape and the sound "wobbles" because the speed fluctuates.

Holzhaus commented:
> Most controllers have only a single sync button that is lit when the deck is either leader or follower. How is the explicit leader indicated?

On my controller I indicate this using a blinking sync button. Yes, blinking is annoying but in this case it actually makes sense IMHO, because usually you don't use explicit leader all the time and it's important information that has been set explicitly by the user.

daschuer commented:
Yes a blinking sync button would make sense for me as well. 

It does however not solve the issue that you don't see if the rate slider is out of sync or not. This would be solved with the #1767 where the rate slider never go out of sync. Or would be an extra indicator or sync state helpful? 

By the way: Thank you for the "Robert Ouimet - Rainbows Refunked" is quite significant for the issue we have here, because there is a slight tempo change in the autro region. The user should definitely have the choice how to deal with it. If you for instance mix the track into itself, the constant clock provides a good results.

ywwg commented:
> When you enable sync on two decks, the rate slider of the first deck is is still snapped in while the rate slider of the second deck is out of sync. With an explicit master, you have an indicator which rate slider is the one that instantly controls the tempo and which need to be snapped in.

The rate sliders are never "out of sync".

Here's a demo with two beatmapped tracks.  You'll see that as I select the leader, the rate slider stops moving around. The tracks are still in sync, but one of them is playing back at a constant playback rate.  Unfortunately the sound didn't record, but you can see the difference in the way the rate sliders behave depending on which is the leader.

https://user-images.githubusercontent.com/888940/111198045-88ff8300-8595-11eb-93f6-88d99c549d02.mov

In the old behavior, or with Internal Clock as Explicit Leader, both rate sliders would be moving around to keep the bpm constant.

ywwg commented:
> Yes a blinking sync button would make sense for me as well.

I'd like to have this PR only be about the engine code, with the UI either temporary as it is, or not even part of this commit.  The UI stuff could be refined out in a later PR.  (As described above, the behavior of Sync Lock will be improved even without the new UI button)

daschuer commented:
> The rate sliders are never "out of sync".

I was referring the rate slider on the controller. They are always out of sync if they are programmatically moved.

This is an issue for any sync strategy.

With an explicitly leader, only the follower rate sliders are out of sync which is an improvement. However we have still no clear indicator for this state.

ywwg committed:
Sync Lock: Fix clazy error

ywwg commented:
Fixed clazy error

Holzhaus commented:
> > The rate sliders are never "out of sync".
> 
> I was referring the rate slider on the controller. They are always out of sync if they are programmatically moved.
> 
> This is an issue for any sync strategy.
> 
> With an explicitly leader, only the follower rate sliders are out of sync which is an improvement. However we have still no clear indicator for this state. 

I agree that this is an issue, but this isn't specific to this PR or Sync. It's an issue with all slider and knobs. For example, my controller mixes main in hardware so the main knob in the GUI is always wrong. Same goes for any SoftTakeover knob/sliders. We discussed a ghost indicator (like we have for spinnies) on Zulip: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/QML/near/229103153

Let's continue that part of the discussion there.

Holzhaus commented:
@ywwg i agree that skin/GUI changes should be left for a follow-up PR. Rebasing or resolving merge conflicts in XML is a nightmare IMHO. And this works fine even without any GUI changes, because in that case explicit leader is just not available to the user.

daschuer commented:
@ywwg would you mind to rebase this PR soon to get rid of the outdated early commits? Than I can start reviewing without bothered by s later rebase.

ywwg commented:
I'm not sure which commits you want me to remove, but I squashed a bunch of the early commits. The diff should be pretty decent

ywwg commented:
Hm, the bad change with the reverse-calculation diff is still there.  I'm not sure how to do this without just squashing the whole diff, is that ok?

ywwg commented:
OK I think I figured out how to amend the commits

Holzhaus commented:
I mostly just do a new commit, then rebase and move it below the commit I want to amend and replace pick with `fixup`.

ywwg commented:
> I mostly just do a new commit, then rebase and move it below the commit I want to amend and replace `pick` with `fixup`.

ah good idea.  I did the rebase first and then did "edit" on the steps, but that requires another rebase afterward to get rid of the extra commit.

daschuer commented:
I don't really mind how this is fixed, just for the record the receipt for a single run:

```
git rebase -i upstream/2.3
```

Than comment the unwanted commits with # 
And replace pick with e at the broken commits.
Save the file.
Fix the build  and 
```
git commit -a --amend 
git rebase --continue
````

ywwg commented:
lmk if the commit history is ok now

Holzhaus commented on src/engine/sync/synccontrol.cpp:
```diff
@@ -316,15 +330,6 @@ void SyncControl::setInstantaneousBpm(double bpm) {
-    // If we're close to the end, and master, disable master so we don't stop
-    // the party.
-    if (isMaster(getSyncMode()) &&
-            fractionalPlaypos > kTrackPositionMasterHandoff) {
```
`kTrackPositionMasterHandoff` is now unused, please remove the variable

ywwg replied:
done

ywwg committed:
Sync Lock: remove unneeded const

Holzhaus commented:
@ywwg Thank you for this PR. Previously, we had two types of leader state with sync lock:
1. **Implicit Leader** - Leader clock was determined automatically
2. **Explicit Leader** - Leader was set manuall by the user

We disabled 2 because it was not working correctly, but I had the impression that this was the plan.
In Mixxx 2.3, we had the following:
1. Single Deck playing -> that deck is implicit leader
2. Two or more decks playing -> Internal clock is implicit leader.

With this PR, I'd expect the basic behavior to stay same, but the additional possibility to mark a deck as explicit leader (in lieu of the Internal Clock).

However, I now alway have a deck marked as explicit leader. Here's an example with a single deck playing. I expected `sync_master` to be 1, not 2:
![Peek 2021-03-16 19-26](https://user-images.githubusercontent.com/1834516/111362358-f80cd280-868e-11eb-9406-2aa2e749f01a.gif)
The explicit leader state sticks with the deck, so if I add a second playing deck, deck one stays leader. This means that the internal clock is basically never used anymore.

Is this a bug, or is this change intended? If the latter is the case, can you elaborate your reasons for it? Or am I misunderstanding this?

Holzhaus commented:
The "soft" (implicit) leader state now sticks with the deck that I first enabled sync lock for, even if I add another deck. Before, the internal clock because leader if multiple decks are playing (IIRC). Is that change intended or is that a bug?

Holzhaus commented:
Here's a video that demonstrates the new behavior. After hitting play on the second deck, I'd expect the Internal Clock to be the sync leader, not channel 1:

![Peek 2021-03-16 21-23](https://user-images.githubusercontent.com/1834516/111375135-33fb6400-869e-11eb-961e-7f4f3290e053.gif)

ywwg committed:
Sync Lock: Add needed svg for testing

ywwg commented:
Your observation is correct.  The new behavior of the system is a little different:

* before: the master syncable controls basebpm (underlying track rate), bpm (current playback rate), and beat distance
* new: the master syncable controls the basebpm, but it only *initializes* the Internal Clock with the bpm and beat distance.

So now, the internal clock is actually always in charge of beat distance, so that the master syncable can scratch and it doesn't get transmitted to the other decks.  The master syncable can also stop and other decks won't stop.  This provides a more predictable experience. 

But note, the Internal Clock can still be made explicit master (no UI for this).

Holzhaus commented:
Thanks for the explanation. From the user's perspective I guess this makes sense.

>> Yes a blinking sync button would make sense for me as well.
>
>I'd like to have this PR only be about the engine code, with the UI either temporary as it is, or not even part of this commit. The UI stuff could be refined out in a later PR. (As described above, the behavior of Sync Lock will be improved even without the new UI button)

We were just discussion if the new "explicit" state can be represented on controllers without a dedicated "Sync Leader" button. We can, so this PR won't hurt usability for controller users.

> Try letting an explicit leader hit the end of a track. Is that the expected behavior for you?

If the explicit leader track ends, it loses the explicit leader state and the other track becomes the new (implicit) leader? Yeah, that makes sense to me.

> situations with beatgrid changes. Loading new tracks, reanalyzing tracks, etc. https://bugs.launchpad.net/mixxx/+bug/1808698 esp

Great, that bug seems to be fixed!

Another case I tested: Two tracks playing with sync lock enabled, track A is leader (I tested with both implicit/explicit leader). Now I right-click the deck and reset the leader's beatgrid. The other deck becomes implicit leader :heavy_check_mark: 

Not that much of an issue, but how is the implicit leader state chosen if two decks are playing? I think explicit leader should create a bias for implicit leader:
Two tracks playing with sync lock enabled, track A is implicit leader. Now set track B as explicit leader. If you now click the Leader button of track B again to disable explicit leader, track A becomes implicit leader. I think track B should stay leader, but implicit. Or is there a reason why track A is becomes leader again?

Apart from the last point, I didn't find issues or unexpected behavior with this. Thanks a ton!

ywwg commented:
> Not that much of an issue, but how is the implicit leader state chosen if two decks are playing? I think explicit leader should create a bias for implicit leader:

Yeah I was unsure about that choice, but my reasoning is, if the user didn't want the deck to be leader any more, they wouldn't have clicked the button.  However it would be pretty easy to make a click go from Explicit to Soft, and then a second click to go from Soft to Off. By design, there's no way to "choose" which one is Soft Leader.

Holzhaus commented on src/engine/sync/basesyncablelistener.cpp:
```diff
@@ -124,22 +124,24 @@ void BaseSyncableListener::setMasterBeatDistance(Syncable* pSource, double beatD
+void BaseSyncableListener::setMasterParams(Syncable* pSource) {
+    const double beatDistance = pSource->getBeatDistance();
+    const double baseBpm = pSource->getBaseBpm();
+    const double bpm = pSource->getBpm() > 0 ? pSource->getBpm() : pSource->getBaseBpm();
```
This assumes the BPM value doesn't change between the comparison and the true case. In my midi clock implementation this will be atomic and may change at any time. Maybe we should just store it in a variable first, and do the comparison on that value?

ywwg replied:
sure

Holzhaus commented on src/test/enginesynctest.cpp:
```diff
@@ -47,13 +49,21 @@ class EngineSyncTest : public MockedEngineBackendTest {
+        if (auto mode = ControlObject::getControl(ConfigKey(group, "sync_mode"))
+                                ->get();
+                mode != SYNC_FOLLOWER) {
+            qWarning() << "expected mode SYNC_FOLLOWER, got" << mode;
```
Are you testing these from cases? Shouldn't this use EXPECT?

Holzhaus replied:
Same for the other warnings in the diff.

ywwg replied:
No, because then the test failures always show up at line 55 or whatever, making it impossible to determine what comparison actually failed.  by doing EXPECT_TRUE(isFollower()) I see the correct failure location.

Holzhaus replied:
Ok

Holzhaus commented on src/test/enginesynctest.cpp:
```diff
@@ -84,54 +94,84 @@ class EngineSyncTest : public MockedEngineBackendTest {
         if (group == m_sGroup1) {
             if (m_pEngineSync->getMaster() != m_pChannel1) {
+                qWarning() << "master pointer is wrong";
+                if (m_pEngineSync->getMaster()) {
+                    qWarning() << "it's " << m_pEngineSync->getMaster()->getGroup();
+                } else {
+                    qWarning() << "it's null";
+                }
```
I suggest to use a ternary operator so will show up in the same log line

ywwg replied:
done

Holzhaus commented on src/test/enginesynctest.cpp:
```diff
@@ -84,54 +94,84 @@ class EngineSyncTest : public MockedEngineBackendTest {
             }
         } else if (group == m_sGroup2) {
             if (m_pEngineSync->getMaster() != m_pChannel2) {
+                qWarning() << "master pointer is wrong";
+                if (m_pEngineSync->getMaster()) {
+                    qWarning() << "it's " << m_pEngineSync->getMaster()->getGroup();
+                } else {
+                    qWarning() << "it's null";
+                }
```
Same here

ywwg committed:
Sync Lock: Clicking an Explicit Master converts to Soft Master. Clicking again converts to Follower

Includes a very clever refactor of requestSyncMode to disentangle the choosing of the master from
the choosing of the master params source. Once again this makes things much cleaner

Holzhaus commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,200 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+        }
+    }
+
+    if (auto unique_syncable = getUniquePlayingSyncable(); unique_syncable) {
+        unique_syncable->notifyOnlyPlayingSyncable();
+    }
```
```suggestion
    const auto* uniqueSyncable = getUniquePlayingSyncable();
    if (uniqueSyncable != nullptr) {
        uniqueSyncable->notifyOnlyPlayingSyncable();
    }
```

ywwg replied:
we're not doing c++17 conditionals? but they are so clean!

Be-ing replied:
I agree that doing assignment inside conditional expressions is a bad practice. It is IMO an misfeature of the language that it compiles.

ywwg replied:
It compiles because it's a specifically-added new feature: https://skebanga.github.io/if-with-initializer/.  As for the danger of assignments in conditionals, the compiler should throw a warning for assignments without associated checks But I'll remove it.

ywwg replied:
(the pointer can't be const because the function is non-const)

Holzhaus replied:
But the lvalue can be const, right?

ywwg replied:
it didn't compile, so shrug

Holzhaus replied:
IIRC for plain pointers the const needs to go behind pointer type. But doesn't Metter, let's just leave it.

ywwg committed:
SyncLock: reduce calls to non-atomic gets

ywwg committed:
Sync Lock: Address notes about test syntax

ywwg committed:
Update src/engine/sync/enginesync.cpp

Remove if-with-initializer

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

ywwg commented:
> However it would be pretty easy to make a click go from Explicit to Soft, and then a second click to go from Soft to Off. By design, there's no way to "choose" which one is Soft Leader.

I have done this, let me know if this makes more sense

Holzhaus reviewed (approved):
Thank you very much. This works fine and I didn't notice any issues.

Do you want to remove the skin changes for now, until we discussed the UI with @ronso0? Maybe we can make the button look nicer or even find a better way to display the leader state (e.g. a lock or star icon on the Sync button?).

@daschuer Are you also planning to review this more? If not, we can merge IMHO.

daschuer commented:
Yes I have planned that.

ywwg commented:
I'll wait for @daschuer 's review and then remove the skin changes

ronso0 commented:
> I'll wait for @daschuer 's review and then remove the skin changes

yeah, let's polish it some other time. (didn't build this branch yet, since I remember using Sync _ever_ )

Holzhaus commented:
@ywwg Sorry, #3734 caused a merge conflict. Please resolve.

daschuer commented on src/engine/sync/synccontrol.cpp:
```diff
@@ -129,7 +127,21 @@ void SyncControl::setSyncMode(SyncMode mode) {
     // requires. Bypass confirmation by using setAndConfirm.
     m_pSyncMode->setAndConfirm(mode);
     m_pSyncEnabled->setAndConfirm(mode != SYNC_NONE);
-    m_pSyncMasterEnabled->setAndConfirm(isMaster(mode));
```
the same is also in InternalClock::setSyncMode(SyncMode mode)

daschuer replied:
Mm I am actually unsure, because a toggle attempt sets it directly to Explicite Master. I am not sure if this is not to much magic if we consider a script user. Please add a comment at the constructor that describes the special case here.

ywwg replied:
sync_master as a light wants to show the type of master: off, soft, or explicit.  But as a button, it only works like this:

* if currently off, requests explicit mode
* if currently soft, requests explicit mode
* if currently explicit, changes to soft mode

so it's still a three state button, but the user can't just toggle through all the settings as usual.

ywwg replied:
> The user offset is lost if you play one deck explicit master enabled with another deck and then switch master to the other deck. Expected behavior: The phase offset adusted by rate_temp_up/down shall remain when switching master

oooo that's tough, let me think about how to handle that

ywwg replied:
This took a lot to fix.  The code was not properly considering user offsets in a bunch of circumstances.  It should work better now.

daschuer commented on src/engine/sync/synccontrol.cpp:
```diff
@@ -129,7 +127,21 @@ void SyncControl::setSyncMode(SyncMode mode) {
+    case SYNC_INVALID:
+    case SYNC_NONE:
+    case SYNC_FOLLOWER:
+        m_pSyncMasterEnabled->setAndConfirm(0);
```
Did you consider to add an enum for these values?

ywwg replied:
ok, and added a conversion func

daschuer reviewed (changes requested):
Thank you for this nice piece. 
I have left some comments  and start manual tests now.

daschuer commented:
Now the soft master behavior is broken that hands over the current BPM to the internal clock. 
I have loaded two non const track in two decks and enabled sync on both. 
I have started to play deck 1, It plays as expected with its normal rate. 
Than I have played the second track. It starts to move the rate slider heavily to follow the first track. 
Expectation: The tempo shall be kept const for both until the transition is over.  
This is the original behavior without an explicit master selected. 

I think @Holzhaus has shown this in: 
https://github.com/mixxxdj/mixxx/pull/3698#issuecomment-800580707

daschuer commented on src/engine/sync/synccontrol.cpp:
```diff
@@ -129,7 +127,21 @@ void SyncControl::setSyncMode(SyncMode mode) {
+        m_pSyncMasterEnabled->setAndConfirm(0);
+        break;
+    case SYNC_MASTER_SOFT:
+        m_pSyncMasterEnabled->setAndConfirm(1);
```
I think m_pSyncMasterEnabled->setStates() needs to be adjusted as well. 
The best these magic numbers shall become an enum.

ywwg replied:
done

daschuer commented:
The user offset is lost if you play one deck explicit master enabled with another deck and then switch master to the other deck. Expected behavior: The phase offset adusted by rate_temp_up/down shall remain when switching master

ywwg committed:
Merge branch 'main' into synclock-explicit-rebased

ywwg committed:
Remove files not needed since merge

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+    // the pSyncable's new mode (it may not be the one they requested),
+    // and activating the appropriate modes in it as well as possibly other
+    // decks that need to change as a result.
+    Syncable* oldMaster = nullptr;
```
oh that should be m_pMasterSyncable

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+            activateMaster(pSyncable, SYNC_MASTER_EXPLICIT);
+        } else {
+            // If we have no bpm, we can't be a valid master. Override.
+            activateFollower(pSyncable);
```
>  Does it mean here, the user pressed the master button on an empty deck and the deck becomes a follower.

yes exactly.

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+        break;
+    }
+    case SYNC_MASTER_SOFT: {
+        if (m_pMasterSyncable == pSyncable &&
```
hah! It turns out the two master cases are completely the same.  Much simpler now

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+        // master mode.
+        if (m_pMasterSyncable == pSyncable) {
+            m_pMasterSyncable = nullptr;
+            activateFollower(pSyncable);
```
Yes you're right.  Did some more rearranging.  Tests still pass...

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+    if (kLogger.traceEnabled()) {
+        kLogger.trace() << "EngineSync::deactivateSync" << pSyncable->getGroup();
+    }
+    bool wasMaster = isMaster(pSyncable->getSyncMode());
```
done

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+
+    if (pSyncable != m_pInternalClock && !bSyncDeckExists) {
+        // Deactivate the internal clock if there are no more sync decks left.
+        m_pMasterSyncable = nullptr;
```
mm yeah

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+
+    Syncable* newMaster = pickMaster(nullptr);
+    if (newMaster != nullptr && m_pMasterSyncable != newMaster) {
+        activateMaster(newMaster, SYNC_MASTER_SOFT);
```
pickmaster will prefer an explicit master, which will cause the m_pmastersyncable pointer not to change, and therefore this code block won't be executed.  I could make this clearer by only calling this code if we're deactivating sync on the master deck.

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+    // parameters, if any. Usually this is the new master. (Note, that pointer might be null!)
+    Syncable* pParamsSyncable = m_pMasterSyncable;
+    // But If we are newly soft master, we need to match to some other deck.
+    if (pSyncable == m_pMasterSyncable && pSyncable != oldMaster && mode != SYNC_MASTER_EXPLICIT) {
```
fixed

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+        }
+    }
+
+    auto* uniqueSyncable = getUniquePlayingSyncable();
```
(applied manually)

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -19,14 +19,201 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+    }
+}
+
+void EngineSync::requestEnableSync(Syncable* pSyncable, bool bEnabled) {
```
yup good idea

ywwg commented on src/engine/sync/enginesync.cpp:
```diff
@@ -253,11 +338,14 @@ void EngineSync::notifyPlaying(Syncable* pSyncable, bool playing) {
+        activateMaster(newMaster, SYNC_MASTER_SOFT);
+    }
+
+    if (auto unique_syncable = getUniquePlayingSyncable(); unique_syncable) {
```
applied manually

ywwg commented:
big batch of comments addressed, more to come.

ywwg committed:
Sync Lock: Address notes from dashuer

ywwg commented:
> Than I have played the second track. It starts to move the rate slider heavily to follow the first track.

This is the new intended behavior.  See Zulip discussion here: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/sync.20lock.20with.20variable.20tempo.20tracks

If the user wants to keep a single locked bpm in that situation, they should activate Internal Clock as explicit master. (There is currently no UI for this but the tests cover that case).

ywwg committed:
Sync Lock: Fix capitalization

daschuer commented:
>> This is the new intended behavior. See Zulip discussion here: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/sync.20lock.20with.20variable.20tempo.20tracks

> If the user wants to keep a single locked bpm in that situation, they should activate Internal Clock as explicit master. (There is currently no UI for this but the tests cover that case).

The issue here is that all tracks follow the BPM of one track WITHOUT explicit selected as leader. 
This is an unintended bug in this branch. Currently we have no difference between explicit and implicit leader.

daschuer commented:
I have tested a bit more, and the on the fly half/double feature "fixed" in https://github.com/mixxxdj/mixxx/pull/3706 seems to be actually a regression in this branch. This can really mess up the mixxx. 
I think this should be fixed here before merge. Can you re-base you changes on top of this? It is currently conflicting.

daschuer commented:
Here the video showing the behavior I am complaining about: 


https://user-images.githubusercontent.com/1777442/112231918-ec458100-8c37-11eb-8405-9a9c15ac6a61.mp4

More details are here: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/sync.20lock.20edge.20cases

ywwg committed:
Sync Lock: use an enum for master light display

ywwg committed:
Sync Lock: make sure sync_master is a three-state control

daschuer commented:
I really like to make progress here.
Unfortunately these PR includes two mayor changes. Only one is announced I the description.

Please split the change of the implicit leader feature, not switching to the internal clock anymore to another PR. This way we have the chance to improve the explicit feature without being cought by the discussiin about the implicit leader change.

ywwg commented:
> Here the video showing the behavior I am complaining about:

what does it look like in 2.3?

daschuer commented:
Here is the same transition using the 2.3 branch: 

https://user-images.githubusercontent.com/1777442/112677467-1088a400-8e6a-11eb-9ce0-58813550012a.mp4

ywwg commented:
> Here is the same transition using the 2.3 branch:

would it be possible to record again with a higher bitrate?  I can't read the bpm values at the beginning of the video.  there's some behavior of the rate sliders that looks odd to me -- this is just straight auto dj right, with no controls being moved by you, right?

ywwg committed:
Sync Lock: Fix user offsets not being maintained across play/pause and master switches

ywwg commented:
> The user offset is lost if you play one deck explicit master enabled with another deck and then switch master to the other deck. Expected behavior: The phase offset adusted by rate_temp_up/down shall remain when switching master

This took a lot to fix. The code was not properly considering user offsets in a bunch of circumstances. It should work better now.

ywwg commented on src/test/enginesynctest.cpp:
```diff
@@ -2146,20 +2146,125 @@ TEST_F(EngineSyncTest, UserTweakPreservedInSeek) {
     // Now, the locations are much more different - but the beat distance appears
     // to be the same because beat distance CO hides the user offset.
-    EXPECT_NEAR(0.2269025,
+    EXPECT_NEAR(0.1990531,
```
I am not 100% sure that the change in these values is right, but I did change the seek phase code so it's plausible that this is right

daschuer replied:
At least it is suspicious. It would be great if we can calc here the expected result. 
For now it is fine for me to test the case manual and it if it is still OK let go for it. 
I think this was the way we have found the constant originally.

daschuer replied:
This was may be wrong initially, In case I seek a track with a user offset in 2.3 branch, the offset is ignored during the seek and than dragged back into phase. This is fixed here. Thank you.

ywwg committed:
Sync Lock: remove redundant comment

ywwg committed:
Merge branch 'main' into synclock-explicit-rebased

daschuer commented:
> would it be possible to record again with a higher bitrate? I can't read the bpm values at the beginning of the video. 

No problem, here it is: 
https://mixxx.zulipchat.com/user_uploads/2380/GqIqBWZ1h7be9iSuDztgGHiC/simplescreenrecorder-2021-03-28_12.28.27.mkv

> there's some behavior of the rate sliders that looks odd to me -- this is just straight auto dj right, with no controls being moved by you, right?

Yes.
* Both tracks are analyzed with non const beat map 
* Loaded into Auto DJ 
* Enable auto DJ 
* latch sync on deck one 
* Latch sync on deck two 
* watch ..

daschuer commented:
>  This PR is at an impasse, I'm not sure how else to resolve the issue.

This PR is at an impasse, because it adds the unrelated change to the implicit leader behavior. The original changes are noncontroversial. How about strip that down to the agreed feature set and postpone the implicit part to a more experimental PR that can be merged after the usability has been proved as discussed.

Be-ing commented:
> This PR is at an impasse, because it adds the unrelated change to the implicit leader behavior.

Please, let's not remove useful features because they create regressions for far edge cases. Solving the root issues requires improving the analyzer, so let's move forward and stop bikeshedding here.

Be-ing commented:
There are merge conflicts now.

ywwg committed:
Merge branch 'main' into synclock-explicit-rebased

ywwg committed:
Sync Lock: Add an option to restore old 2.3 leader selection behavior

Add the option:

[BPM]
SyncLockPickAlgorithm 1

To use the 2.3 leader selection behavior, which defaults to the internal clock when
more than one sync locked deck is playing.

ywwg commented:
synced and added a config option to use the old 2.3 leader selection behavior. 

add:

```
[BPM]
    SyncLockPickAlgorithm 1
```
to mixxx.cfg to use it

@daschuer

daschuer commented:
Did you mean:
SyncLockPickClockAlgorithm?
Or just a more generic
SyncLockAlgorithm?

daschuer commented:
But I don't mind since it is temporary.
Thank you for making this real.

Holzhaus commented:
Thanks. I think we decided that we use `snake_case` for new COs/Config options. But since it's temporary IDC.

ywwg committed:
Sync Lock: tweak the names

ywwg commented:
made the small changes, SyncLockAlgorithm and sync_lock_algorithm

Holzhaus commented:
Some merge conflicts have evolved. Can you fix them? Then I'll give this a final test and we can merge IMHO.

daschuer commented:
Is the nudging issue solved? I like to give this also a final test.

ywwg commented:
I am trying to fix the merge conflicts but now the tests fail :(.  Trying to figure out what happened, I think it has something to do with the Audible changes

ywwg committed:
Merge branch 'main' into synclock-explicit-rebased

ywwg commented:
@daschuer I think your audible patch is broken -- it only sets the audible flag on PostProcess, but some of the tests result in calls to pickMaster before any buffers have been processed.  Therefore the synccontrol doesn't think it's audible when it actually is.  I have tried to add ProcessBuffer calls but it's not working.  I really wish you would have committed this PR first before making your change.  I don't know when I'll have time to figure out why these tests aren't working.

Holzhaus committed:
Translations: Print warning if translations failed to load

Holzhaus committed:
Translations: Improve locale override logic

Holzhaus committed:
Translations: Use helper method for unified error handling

Holzhaus committed:
Translations: Simplify code (remove scope and redundant default constrs)

Holzhaus committed:
Translations: Use default value when reading from config

Holzhaus committed:
Translations: Use Debug instead of Warning loglevel for mixxx qt strings

Holzhaus committed:
res/translations: Delete outdates/unused "ceb" translations

See https://github.com/mixxxdj/mixxx/pull/3826#issuecomment-834334745.

uklotzde committed:
Import metadata and cover image without temporary track object

ywwg committed:
Sync Lock: fix missing checks for playing && audible

ywwg committed:
Sync Lock: Fix tests

ywwg commented:
fixed tests!

daschuer commented:
Greate news. Thank you very much.

daschuer commented:
At least we can add this review to our dated press mirror
https://mixxx.org/press/

Owen Williams ***@***.***> schrieb am So., 9. Mai 2021, 05:41:

> fixed tests!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/3698#issuecomment-835659120>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANR6IQR75RUE4XENEA2RG3TMX77NANCNFSM4ZC3ZMVA>
> .
>

ywwg commented:
any thing else to do before merge?

daschuer commented:
There is a wierd pre-commit issue. I will restart it.

daschuer reviewed (changes requested):
The CI issue is gone now. 

Unfortunately I have found one issue that prevent merging: 
The momentary sync feature (short press on sync) does no longer sync the tempo of a pause track to to the playing track when sync_lock_algorithm = 1. With sync_lock_algorithm = 0 it is working.  

Than we have the known issues, that the instant double feature does not copy the user offset (also an issue in 2.3)  and the repeated recalculation for the double/half feature during sync in sync_lock_algorithm = 0 mode, targeted in https://github.com/mixxxdj/mixxx/pull/3706. They can be fixed later.

Holzhaus committed:
Translations: Use POSIX-style locale identifiers

Switch to using an underscore to separate ISO-639-1 language and
ISO-3166-1 country codes (instead of a dash).

Also removes the `lang_map` setting from the Transifex client
configuration that replaces the underscore with a dash:
https://docs.transifex.com/client/client-configuration#language-mappings

This fixes the broken loading of translations that used a dash in the
filename, because this is not supported by `QTranslator::load()`.
See https://doc.qt.io/qt-5/qtranslator.html#load-1 for details.

uklotzde committed:
Track metadata import/update: Rename enumeration

uklotzde committed:
Indicate if modified/unmodified when importing track metadata

ywwg committed:
Sync Lock: Fix missing case in prefer-internal-clock

ywwg commented:
@daschuer fixed the case you mentioned and added a test

daschuer commented:
I can confirm that the remaining issue is fixed. 
Unfortunately #3854 has slipped in here along with some unnecessary merges from main.  

Can you rebase it out? 
I think it is 
```
rebase --onto db06e5ac4896e8882fe9fe139262ee5ad3429579 79fced599fd3523b772bfe2ebe8880181e2f5aa2
```

Thank you.

ywwg commented:
rebased

ywwg committed:
Merge remote-tracking branch 'mixxx/main' into synclock-explicit-rebased

ywwg commented:
fixed merge conflict I think

daschuer commented:
Now Jan's commits slipped in.

I have merged this without these commits in 6ed6a095cbc44c7bc7c7afbbef4752ffda627e0d

daschuer commented:
Thank you for bringing this controversial PR to a mergable state.  This will make a big difference once the beat map editor is in place.

Holzhaus commented:
One thing I noticed: When the track ends, it's still explicit leader. Is that intended? I don't think that makes much sense, unless we want to restore the "scratch two decks at the same time".

Holzhaus commented:
I can even set explicit leader on a stopped deck, which changes the key on the playing deck.

daschuer commented:
All these questions are still floating IMHO, bound to the question how the user expect to use the single sync button and single sync LED on his controller and how the internal clock is expected to be controlled.

For a good decision, I think we need the whole picture how other tools get around these problems. 

The idea was to set up a wiki page and describe how the sync button works in other tools with all available preferences option. This shows us the expectations user migrating to Mixxx might have.

Than we need to step back and understand the use cases covered and look if and how Mixxx schould cover them.

My hope is that we can allow a per transition decision in the GUI without any prefernces option.

For now, I like to allow these modes:
* Smart mode where Mixxx decides 
* Allow to "help" Mixxx if it is not smart enough
* Hard override Smart mode, including scratching.

I have no idea yet how to control it effective. That's why a market research might be helpful. At least that may help that new users don't need to learn completely new.  

If you agree I can start such a wiki page.

Holzhaus commented:
> All these questions are still floating IMHO, bound to the question how the user expect to use the single sync button and single sync LED on his controller and how the internal clock is expected to be controlled


I think the explicit leader behavior I reported above is a bug. I'm pretty certain that last time I tested this branch did reset the explicit leader state when the end of track is reached.

Once a track stopped it is not the explicit leader anymore (there is no beatgrid, so I assume the internal clock takes over?)
The only use case for setting/maintaining explicit leader on a stopped I can think of is that you can use it to scratch both decks at once, which is something that is deliberately not covered by this branch because it's an extreme edge case and I think @ywwg left that out because we should streamline the usual workflow first and iterate based on that.

daschuer commented:
Is this only a visual issue, or is a use case affected?

Holzhaus commented:
Both. It's a visual issue, but the explicit leader decks stays leader forever. The next track I load on that deck will also be leader.

And also this:

> I can even set explicit leader on a stopped deck, which changes the key on the playing deck.

There is an audible change on the playing deck. No idea why, because the tempo does not change.

ywwg commented:
Tracks should lose explicit leader at track end. I'll fix

ywwg commented:
it's ok for an explicit leader to be stopped! That is definitely intended.  It's still driving the bpm, but it doesn't need to be playing
